### DISCUSSION TITLE
Midi overrun with controller messages?

### DISCUSSION THREAD

**Discussion by jawheaton** [FortySevenEffects/arduino_midi_library#293] [General]
Hello - 
First thanks so much for the work on this library!
I have been using the library with an Arduino Mega 2560, and the Adafruit MIDI featherwing midi interface (https://www.adafruit.com/product/4740) .  I am really only interested in reading MIDI and controlling some lights based on the MIDI input.

I have created callbacks for noteon noteoff and pitchchange.  In general things are very responsive (low-latency) and accurate.  

My problem is that when control change messages are sent I get bad behavior in my code.  It seems to miss note-off messages and interprets other CC messages as note ons.  It doesn't happen with pitch-bend, so I'm not sure if it is a hardware or serial port issue or if it is something in the library.

Of course, I will do some more debugging, but that also adds latency to the system so my question is does anyone have this same experience?

For code, here are the important things I am doing:

      MIDI_CREATE_INSTANCE(HardwareSerial, Serial3, midi2);
     ...
      Serial3.begin(31250);
      midi2.setHandleNoteOn(handleNoteOn);  
      midi2.setHandleNoteOff(handleNoteOff);
      midi2.setHandlePitchBend(handlePitchBend); 
      midi2.begin(MIDI_CHANNEL_OMNI);      // Launch MIDI and listen to all channels
     ...
    if (midi2.read())                    // If we have received a message
    {
        digitalWrite(LED_BUILTIN, HIGH);
        digitalWrite(LED_BUILTIN, LOW);
        if (gDebug & DEBUG_MIDI)
        {
          //Serial.print("MIDI MESSAGE: "); Serial.println(midi2.getType());
        }
    }

**Comment by jawheaton** [comment]
As an update, the problem almost completely goes away if I do NOT assign the call back functions.  In other words if I use the raw midi message data after the read() and process them myself, I do not have the issues of the control change messages doing bad things.  But if I really beat on the keybaord while  twiddling a controller knob I can sometimes get invalid midi data (like values > 127 for the 2nd data byte).  So I'm going to proceed with this method.

**Reply by franky47** [reply]
Interesting, do you have a way to record the MIDI stream between the keyboard and the Arduino?

I would have suggested RX buffer overflows, which would drop data and maybe skip some messages (which can happen if the main loop spins too slowly), but that wouldn't explain invalid data bytes.

**Comment by franky47** [comment]
As a side note, the line `Serial3.begin(31250);` is unnecessary, `MIDI.begin` will deal with transport initialisation.

**Reply by jawheaton** [reply]
Thanks for the feedback.  I have removed serial port init statement.  At first I thought that made a difference, but really I think there is another problem.  I have a repeatable condition (sort of).  I am using a midi keyboard controller as the input device.  If I hold a single key note down and rapidly move both the pitch and modulation (cc 1) wheels back and forth, I sometimes miss the note off, and sometimes get double note on message (without a note off in between) and sometimes get note on messages with bad values (velocity > 127), etc... 

However,  I tried a second kind of midi keyboard controller and I don't get the problem.  SO I THINK MY CHEAP OLD keyboard controller was actually sending bad data when it was over loaded with commands!  (Garbage In Garbage Out).

Next up is some testing with more modern synths!

**Reply by franky47** [reply]
I see, thanks for reporting back.

However, the MIDI parser should be resilient to bad data and just drop invalid messages (so you should never have data bytes > 127). A dump of that invalid stream would be very helpful for us in making sure the parser is airtight.

**Reply by jawheaton** [reply]
Is there an easy way to get the dump of the data from within the library?  (Or do I need to print the raw bytes to a file somehow?) 

On another note, do you think that Use1ByteParsing would affect things?  I think it is defaulted to true at the moment.

**Reply by jawheaton** [reply]
I looks like I can force the errors to happen when Use1ByteParsing is true, but when I set it to false, I don't see the problems (but it is hard to get a completely reproducible data).  I force the error by hitting lots of keys up and down the keyboard while moving both pitch and modulation wheel rapidly.   I know that is not very likely in "real life", but my device has been known to miss midi messages in recording sessions when players are using a lot of controller messages (like filter sweep knobs).

**Reply by franky47** [reply]
> Is there an easy way to get the dump of the data from within the library? (Or do I need to print the raw bytes to a file somehow?)

Not from the library no.

The Use1ByteParsing option is here to alleviate problems in sketches with slow-running main loops, that may otherwise process incoming data with latency: setting it to `false` parses everything in one go, and the use of callbacks is highly recommended (as only the last received message is available when leaving MIDI.read).

When it's set to `true` and a lot of data comes into a slow-running main loop, it may occur that the RX buffer overflows, drop some data, and the next MIDI.read will pick up from later on in the stream. It's probably what's happening here. You could try and increase the RX buffer size to see if it solves the problem.

**Comment by jhaimerl1** [comment]
I just posted an issue I'm having a few minutes ago. My program also misses note-on and off messages. However, this only seems to be happening when I have the the MIDI Clock enabled. I'm curious if you are testing it with clock on or off...

**Reply by jawheaton** [reply]
I am not using MIDI clock.  That is to say, if MIDI clock is being sent, I ignore it.  My only purpose is to read the incoming note on/off information and controllers/pitch bend.  In my current setting, I am setting Use1ByteParsing to true and NOT setting any of the message handlers.  I know this isn't what was recommended by @franky47, but I am empirically getting better results, so I'm not going to mess with it!

Also, there was a recommendation to increase the RX buffer size, but I haven't looked in to how to do that yet.

**Reply by jhaimerl1** [reply]
I tested my setup just now at up to 300 BPM tempo from abeton live. Everything works great! As soon as I turn on the synchronization button in preferences for my interface, and clock messages start being sent, I have data loss.  I looked at my USB MIDI interface that I bought to test my setup... While the reviews are about 4.5 stars over thousands of units, some of the negative 1-star reviews  talk about dropped note on & off messages when time synch is used!  I just ordered a Roland replacement and have another unit sitting over at my drummer's house in the mixer case.  Keeping my fingers crossed that it is this piece of crap and not my circuit or the code!

**Reply by franky47** [reply]
Well if the code is the same and it works well on one hardware and not well on another, I'd wager it's not the code 😄

**Comment by jawheaton** [comment]
I had a similar issue where my really old keyboard controller would put out bad data when it was being taxed.   A newer model didn’t have that issue!
Jim

**Comment by jawheaton** [comment]
I was still having dropped note messages (as far as my code was concerned), so I followed instructions from this link (https://www.hobbytronics.co.uk/arduino-serial-buffer-size)  (modified to find the correct location of the sketch folder on my windows machine).  I was able to increase the RX and TX buffer sizes from 64 to 256 bytes and now I have trouble forcing any lost messages.  My main arduino loop (turning on 60 LEDs takes about 6 mSecs to complete, which is about 6 3-byte midi messages at full baudrate.  So I can see where if I get lots of CC messages very quickly, it could overrun my input buffer.  Fingers crossed!
